src/utils/contracts/AdapterInvestLimitBase.sol

```diff
@@ -0,0 +1,113 @@
+    }
+
+    /**
+     * @inheritdoc IAdapterInvestLimit
+     */
+    function setMaxDepositPoolPct(address _liquidityPool, uint256 _maxDepositPoolPct)
+        external
+        override
+        onlyRiskOperator
+    {
+        maxDepositPoolPct[_liquidityPool] = _maxDepositPoolPct;
+        emit LogMaxDepositPoolPct(maxDepositPoolPct[_liquidityPool], msg.sender);
+    }
+
+    /**
+     * @inheritdoc IAdapterInvestLimit
+     */
+    function setMaxDepositAmount(
+        address _liquidityPool,
+        address _underlyingToken,
+        uint256 _maxDepositAmount
+    ) external override onlyRiskOperator {
+        maxDepositAmount[_liquidityPool][_underlyingToken] = _maxDepositAmount;
+        emit LogMaxDepositAmount(maxDepositAmount[_liquidityPool][_underlyingToken], msg.sender);
+    }
+
+    /**
+     * @inheritdoc IAdapterInvestLimit
+     */
+    function setMaxDepositProtocolMode(MaxExposure _mode) external override onlyRiskOperator {
+        maxDepositProtocolMode = _mode;
+        emit LogMaxDepositProtocolMode(maxDepositProtocolMode, msg.sender);
+    }
+
+    /**
+     * @inheritdoc IAdapterInvestLimit
+     */
+    function setMaxDepositProtocolPct(uint256 _maxDepositProtocolPct) external override onlyRiskOperator {
+        maxDepositProtocolPct = _maxDepositProtocolPct;
+        emit LogMaxDepositProtocolPct(maxDepositProtocolPct, msg.sender);
+    }
+
+    /**
+     * @dev Get the final value of amount in underlying token to be deposited
+     * @param _liquidityPool liquidity pool address
+     * @param _underlyingToken underlying token address
+     * @param _amount amount in underlying token
+     * @param _poolValue pool value
+     * @return amount in underlying token to be deposited affected by investment limitation
+     */
+    function _getDepositAmount(
+        address _liquidityPool,
+        address _underlyingToken,
+        uint256 _amount,
+        uint256 _poolValue
+    ) internal view returns (uint256) {
+        uint256 _limit =
+            maxDepositProtocolMode == MaxExposure.Pct
+                ? _getMaxDepositAmountByPct(_liquidityPool, _poolValue)
+                : maxDepositAmount[_liquidityPool][_underlyingToken];
+        return _amount > _limit ? _limit : _amount;
+    }
+
+    /**
+     * @dev Gets the maximum amount in underlying token limited by percentage
+     * @param _liquidityPool liquidity pool address
+     * @param _poolValue pool value
+     * @return  amount in underlying token to be deposited affected by
+     *          investment limit in percentage
+     */
+    function _getMaxDepositAmountByPct(address _liquidityPool, uint256 _poolValue) internal view returns (uint256) {
+        uint256 _poolPct = maxDepositPoolPct[_liquidityPool];
+        uint256 _limit =
+            _poolPct == 0
+                ? (_poolValue.mul(maxDepositProtocolPct)).div(uint256(10000))
+                : (_poolValue.mul(_poolPct)).div(uint256(10000));
+        return _limit;
+    }
```

Comment:
- [ ] If `_getMaxDepositAmountByPct` function is made public and extended as a part of the `IAdapterInvestLimitV2` interface Vault can benefit our of it by making sure deposit and withdraw actually happened.

(Note : - There are protocols like Compound and forks of Compound which return some error value on failure while doing deposit/withdraw. In this scenario, pre calculating the amount that will be deposited to protocol will ensure robustness of `rebalance` and `vaultDepositAllToStrategy` functions)

While performing investment limit calculation, the PoolValue should be subtracted with `allAmountInToken ` of the vault. Imagine recursively calling `rebalance` / `vaultDepositAllToStrategy` , it will keep on depositing without considering previous deposits that vault made. When `getDepositXCodes()` calls the function, it should pass PoolValue minus allAmountInToken